# Import libraries

In [92]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [93]:
matches = pd.read_csv('data/matches.csv')
deliveries = pd.read_csv('data/deliveries.csv')

In [94]:
deliveries.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [95]:
matches.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

#### Lets look at the shape of each dataframes

In [96]:
deliveries.shape

(179078, 21)

In [97]:
matches.shape

(756, 18)

#### Merge the two dataframes
I used inner join to merged the two dataframes.

In [98]:
data = matches.merge(deliveries, left_on='id', right_on='match_id')

In [99]:
data.head()

id  season       city        date                team1  \
0   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   
1   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   
2   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   
3   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   
4   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
2  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
3  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
4  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

   result  dl_applied  ... bye_runs  legbye_runs  noball_runs penalty_runs  \
0  normal           0  ...        0            0            0            0   
1  normal           0  ...        0            0            0            0   
2  normal           0  ...        0            0            0            0   
3  normal           0  ...        0            0            0            0   
4  normal           0  ...        0            0            0            0   

  batsman_runs extra_runs total_runs player_dismissed  dismissal_kind  fielder  
0            0          0          0              NaN             NaN      NaN  
1            0          0          0              NaN             NaN      NaN  
2            4          0          4              NaN             NaN      NaN  
3            0          0          0              NaN             NaN      NaN  
4            0          2          2              NaN             NaN      NaN  

[5 rows x 39 columns]

#### Lets look at the size of the resulting dataframe and the data types of each column.

In [100]:
data.dtypes

id                   int64
season               int64
city                object
date                object
team1               object
team2               object
toss_winner         object
toss_decision       object
result              object
dl_applied           int64
winner              object
win_by_runs          int64
win_by_wickets       int64
player_of_match     object
venue               object
umpire1             object
umpire2             object
umpire3             object
match_id             int64
inning               int64
batting_team        object
bowling_team        object
over                 int64
ball                 int64
batsman             object
non_striker         object
bowler              object
is_super_over        int64
wide_runs            int64
bye_runs             int64
legbye_runs          int64
noball_runs          int64
penalty_runs         int64
batsman_runs         int64
extra_runs           int64
total_runs           int64
player_dismissed    object
d

Seems all the data types are good.

**To do:**
1. Convert date to datetime

In [101]:
data.shape

(179078, 39)

After merging the two dataframes, the resulting dataframe has the same size that of the deliveries dataframe.

#### Counting the number null values in each column, we have

#### 

In [102]:
data.isnull().sum().sort_values(ascending=False)

fielder             172630
player_dismissed    170244
dismissal_kind      170244
umpire3             150712
city                  1700
umpire2                500
umpire1                500
player_of_match        372
winner                 372
season                   0
win_by_runs              0
venue                    0
win_by_wickets           0
dl_applied               0
toss_decision            0
toss_winner              0
team2                    0
team1                    0
date                     0
result                   0
inning                   0
match_id                 0
batting_team             0
total_runs               0
extra_runs               0
batsman_runs             0
penalty_runs             0
noball_runs              0
legbye_runs              0
bye_runs                 0
wide_runs                0
is_super_over            0
bowler                   0
non_striker              0
batsman                  0
ball                     0
over                     0
b

There's a lot of missing values from umpire3, player_dismissed, dismissal_kind and fielder. It's normal to have no players being ejected from a match. And some matches don't have neccesarily need three umpires. So we can leave the null values on those columns as is.

**To do:**

1. How to deal with missing values in city?
2. There are 372 columns with no winner and players of the match indicated, how to deal with this?
3. There are 500 rows with no umpire1 and umpire2. Given those, is it an official match? What can we do with this?

### Most successful teams
I decided to label the most successful team with the highest number of wins.

In [103]:
# count the win by each team
sucTeam = matches['winner'].value_counts().to_frame()

# rename column
sucTeam.rename(columns={'winner':'Count of Wins'}, inplace=True)
sucTeam.head()

Count of Wins
Mumbai Indians                         109
Chennai Super Kings                    100
Kolkata Knight Riders                   92
Royal Challengers Bangalore             84
Kings XI Punjab                         82

In [104]:
print('The most successful team is {}.'.format(sucTeam.index[0]))

The most successful team is Mumbai Indians.


### Most successful players 
Count the how many times a player became the player of the match. The players with most number of times that become the player of the match are the most successful ones.

In [105]:
# count how many times the player become the player of the match
playerOfMatch = matches['player_of_match'].value_counts().to_frame()

# rename column
playerOfMatch.rename(columns={'player_of_match':'Player of the Match Count'}, inplace=True)
playerOfMatch.head()

Player of the Match Count
CH Gayle                               21
AB de Villiers                         20
DA Warner                              17
RG Sharma                              17
MS Dhoni                               17

In [106]:
print('The most successful player is {}'.format(playerOfMatch.index[0]))

The most successful player is CH Gayle


### Factors contributing win or loss of a team.

**To do:**

1. Does being in the homecourt have an advantage?
2. Does winning the toss coin has impact on the result of the match?

In [107]:
batsmanGrp = deliveries.groupby(["match_id", "inning", "batting_team", "batsman"])
batsmen = batsmanGrp["batsman_runs"].sum().reset_index()
batsmen

match_id  inning         batting_team       batsman  batsman_runs
0             1       1  Sunrisers Hyderabad   BCJ Cutting            16
1             1       1  Sunrisers Hyderabad     DA Warner            14
2             1       1  Sunrisers Hyderabad      DJ Hooda            16
3             1       1  Sunrisers Hyderabad  MC Henriques            52
4             1       1  Sunrisers Hyderabad      S Dhawan            40
...         ...     ...                  ...           ...           ...
11330     11415       2  Chennai Super Kings      MS Dhoni             2
11331     11415       2  Chennai Super Kings     RA Jadeja             9
11332     11415       2  Chennai Super Kings      SK Raina             9
11333     11415       2  Chennai Super Kings     SN Thakur             2
11334     11415       2  Chennai Super Kings     SR Watson            84

[11335 rows x 5 columns]